### PR TITLE
[#195] Abstract the way kernel loads config

### PR DESCRIPTION
**Pull request by AlexanderC**
Abstract the way kernel loads config and implement async config loader.

### CODE REVIEW

File: src/deep-kernel/lib/Config/Loader.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    return this;
+  }
+
+  /**
+   * @param {Function} onLoaded
+   * @param {Function} onFail
+   */
+  load(onLoaded = () => {}, onFail = () => {}) {
+    onLoaded && this._driver.onLoadedCb(onLoaded);
```
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Loader.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201397552/issues/source?bid=2895475&fileBranchId=3169716#l54)

File: src/deep-kernel/lib/Config/Loader.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    onFail && this._driver.onFailCb(onFail);
```
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Loader.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201397552/issues/source?bid=2895475&fileBranchId=3169716#l55)

File: src/deep-kernel/lib/Config/Driver/AsyncConfig.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  /**
+   * @param {String} bucket
+   * @param {String} configFile
+   * @private
+   */
+  _load(bucket = null, configFile = null) {
+    this._bucket = bucket || this._bucket;
+    this._configFile = configFile || this._configFile;
+
+    this._bucket ? this._loadFromS3() : this._loadFromEndpoint();
```
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Driver/AsyncConfig.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201397546/issues/source?bid=2895475&fileBranchId=3169716#l53)

File: src/deep-kernel/lib/Config/Driver/ScopeDriver.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  /**
+   * @param {String} key
+   * @private
+   */
+  _load(key = null) {
+    this._key = key || this._key;
+
+    this._scope.hasOwnProperty(this._key) ?
+      this.loaded(this._scope[this._key]) :
+      this.fail(`Unable to load configuration from scope: No such key '${this._key}' exists`);
```
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Driver/ScopeDriver.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201397551/issues/source?bid=2895475&fileBranchId=3169716#l63)

Comment (devs-deep) [inline src/deep-kernel/lib/Config/Driver/ScopeDriver.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201458612/issues/source?bid=2895475&fileBranchId=3169716#l63)

File: src/deep-kernel/lib/Config/Loader.js (hunk 1 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Loader.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201458613/issues/source?bid=2895475&fileBranchId=3169716#l54)

File: src/deep-kernel/lib/Config/Loader.js (hunk 2 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Loader.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201458613/issues/source?bid=2895475&fileBranchId=3169716#l55)

File: src/deep-kernel/lib/Config/Driver/AsyncConfig.js (hunk 3 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Driver/AsyncConfig.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201458607/issues/source?bid=2895475&fileBranchId=3169716#l53)

File: src/deep-kernel/lib/Config/Driver/ScopeDriver.js (hunk 4 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Driver/ScopeDriver.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201549673/issues/source?bid=2895475&fileBranchId=3169716#l63)

File: src/deep-kernel/lib/Config/Loader.js (hunk 1 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Loader.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201549674/issues/source?bid=2895475&fileBranchId=3169716#l54)

File: src/deep-kernel/lib/Config/Loader.js (hunk 2 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Loader.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201549674/issues/source?bid=2895475&fileBranchId=3169716#l55)

File: src/deep-kernel/lib/Config/Driver/AsyncConfig.js (hunk 3 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Driver/AsyncConfig.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201549668/issues/source?bid=2895475&fileBranchId=3169716#l53)

File: src/deep-kernel/lib/Config/Loader.js (hunk 1 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Loader.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201692602/issues/source?bid=2895475&fileBranchId=3169716#l54)

File: src/deep-kernel/lib/Config/Loader.js (hunk 2 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Loader.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201692602/issues/source?bid=2895475&fileBranchId=3169716#l55)

File: src/deep-kernel/lib/Config/Driver/AsyncConfig.js (hunk 3 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Driver/AsyncConfig.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201692596/issues/source?bid=2895475&fileBranchId=3169716#l53)

File: src/deep-kernel/lib/Config/Driver/ScopeDriver.js (hunk 4 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Driver/ScopeDriver.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201692601/issues/source?bid=2895475&fileBranchId=3169716#l63)

Comment (devs-deep) [inline src/deep-kernel/lib/Config/Driver/ScopeDriver.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201829429/issues/source?bid=2895475&fileBranchId=3169716#l63)

File: src/deep-kernel/lib/Config/Loader.js (hunk 1 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Loader.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201829430/issues/source?bid=2895475&fileBranchId=3169716#l54)

File: src/deep-kernel/lib/Config/Loader.js (hunk 2 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Loader.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201829430/issues/source?bid=2895475&fileBranchId=3169716#l55)

File: src/deep-kernel/lib/Config/Driver/AsyncConfig.js (hunk 3 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Driver/AsyncConfig.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201829424/issues/source?bid=2895475&fileBranchId=3169716#l53)

File: src/deep-kernel/lib/Config/Loader.js (hunk 1 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Loader.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201880697/issues/source?bid=2895475&fileBranchId=3169716#l54)

File: src/deep-kernel/lib/Config/Loader.js (hunk 2 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Loader.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201880697/issues/source?bid=2895475&fileBranchId=3169716#l55)

File: src/deep-kernel/lib/Config/Driver/AsyncConfig.js (hunk 3 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Driver/AsyncConfig.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201880691/issues/source?bid=2895475&fileBranchId=3169716#l53)

File: src/deep-kernel/lib/Config/Driver/ScopeDriver.js (hunk 4 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Driver/ScopeDriver.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2201880696/issues/source?bid=2895475&fileBranchId=3169716#l63)

File: src/deep-kernel/lib/Config/Loader.js (hunk 1 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Loader.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2202795606/issues/source?bid=2895475&fileBranchId=3169716#l54)

File: src/deep-kernel/lib/Config/Loader.js (hunk 2 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Loader.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2202795606/issues/source?bid=2895475&fileBranchId=3169716#l55)

File: src/deep-kernel/lib/Config/Driver/AsyncConfig.js (hunk 3 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Driver/AsyncConfig.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2202795600/issues/source?bid=2895475&fileBranchId=3169716#l53)

File: src/deep-kernel/lib/Config/Driver/ScopeDriver.js (hunk 4 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Driver/ScopeDriver.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2202795605/issues/source?bid=2895475&fileBranchId=3169716#l63)

File: src/deep-kernel/lib/Config/Driver/AsyncConfig.js (hunk 3 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Driver/AsyncConfig.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2202947102/issues/source?bid=2895475&fileBranchId=3169716#l53)

File: src/deep-kernel/lib/Config/Loader.js (hunk 1 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Loader.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2202947108/issues/source?bid=2895475&fileBranchId=3169716#l54)

File: src/deep-kernel/lib/Config/Loader.js (hunk 2 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Loader.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2202947108/issues/source?bid=2895475&fileBranchId=3169716#l55)

File: src/deep-kernel/lib/Config/Driver/ScopeDriver.js (hunk 4 above)
Comment (devs-deep) [inline src/deep-kernel/lib/Config/Driver/ScopeDriver.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/MitocGroup/deep-framework/file/2202947107/issues/source?bid=2895475&fileBranchId=3169716#l63)

**Comment by mgoria** [comment]
Agreed with @AlexanderC to refactor FS implementation by creating a shared namespace into system bucket

**Pull request closed without merging**